// @vue-storefront/magento-api/lib/index.server.js
var Fo=Object.create;var h=Object.defineProperty,Vo=Object.defineProperties,Go=Object.getOwnPropertyDescriptor,Bo=Object.getOwnPropertyDescriptors,zo=Object.getOwnPropertyNames,v=Object.getOwnPropertySymbols,Ho=Object.getPrototypeOf,x=Object.prototype.hasOwnProperty,_e=Object.prototype.propertyIsEnumerable;var pe=(e,r,o)=>r in e?h(e,r,{enumerable:!0,configurable:!0,writable:!0,value:o}):e[r]=o,n=(e,r)=>{for(var o in r||(r={}))x.call(r,o)&&pe(e,o,r[o]);if(v)for(var o of v(r))_e.call(r,o)&&pe(e,o,r[o]);return e},y=(e,r)=>Vo(e,Bo(r)),ye=e=>h(e,"__esModule",{value:!0});var f=(e,r)=>{var o={};for(var t in e)x.call(e,t)&&r.indexOf(t)<0&&(o[t]=e[t]);if(e!=null&&v)for(var t of v(e))r.indexOf(t)<0&&_e.call(e,t)&&(o[t]=e[t]);return o};var fe=(e,r)=>{for(var o in r)h(e,o,{get:r[o],enumerable:!0})},ge=(e,r,o,t)=>{if(r&&typeof r=="object"||typeof r=="function")for(let i of zo(r))!x.call(e,i)&&(o||i!=="default")&&h(e,i,{get:()=>r[i],enumerable:!(t=Go(r,i))||t.enumerable});return e},s=(e,r)=>ge(ye(h(e!=null?Fo(Ho(e)):{},"default",!r&&e&&e.__esModule?{get:()=>e.default,enumerable:!0}:{value:e,enumerable:!0})),e),Do=(e=>(r,o)=>e&&e.get(r)||(o=ge(ye({}),r,1),e&&e.set(r,o),o))(typeof WeakMap!="undefined"?new WeakMap:0);var fa={};fe(fa,{createApiClient:()=>ya,init:()=>Oo});var Lo=require("@vue-storefront/core");var le={};fe(le,{addBundleProductsToCart:()=>he,addConfigurableProductsToCart:()=>Q,addDownloadableProductsToCart:()=>q,addProductToWishList:()=>Ie,addProductsToCart:()=>w,addSimpleProductsToCart:()=>Me,addVirtualProductsToCart:()=>I,applyCouponToCart:()=>A,availableStores:()=>Ve,cart:()=>k,cartTotalQty:()=>De,categoryList:()=>M,categorySearch:()=>R,changeCustomerPassword:()=>Ke,cmsBlocks:()=>S,cmsPage:()=>$,countries:()=>L,country:()=>E,createCustomer:()=>lt,createCustomerAddress:()=>O,createEmptyCart:()=>_t,createProductReview:()=>Ct,currency:()=>F,customMutation:()=>Tt,customQuery:()=>Lt,customer:()=>xt,customerCart:()=>V,customerOrders:()=>At,customerProductReview:()=>Rt,deleteCustomerAddress:()=>G,generateCustomerToken:()=>Bt,getAvailableCustomerPaymentMethods:()=>Dt,getAvailableCustomerShippingMethods:()=>B,getAvailablePaymentMethods:()=>z,getAvailableShippingMethods:()=>Kt,getCustomerAddresses:()=>H,mergeCarts:()=>rr,placeOrder:()=>D,productDetail:()=>N,productReview:()=>lr,productReviewRatingsMetadata:()=>cr,products:()=>W,relatedProduct:()=>Y,removeCouponFromCart:()=>J,removeItemFromCart:()=>j,removeProductsFromWishlist:()=>xr,requestPasswordResetEmail:()=>K,resetPassword:()=>X,revokeCustomerToken:()=>$r,route:()=>ne,setBillingAddressOnCart:()=>Or,setGuestEmailOnCart:()=>Z,setPaymentMethodOnCart:()=>ee,setShippingAddressesOnCart:()=>te,setShippingMethodsOnCart:()=>re,storeConfig:()=>oe,subscribeEmailToNewsletter:()=>ae,updateCartItems:()=>ie,updateCustomer:()=>eo,updateCustomerAddress:()=>ue,updateCustomerEmail:()=>io,upsellProduct:()=>no,urlResolver:()=>se,wishlist:()=>fo,wishlistItemsCount:()=>bo});var Ce=s(require("graphql-tag")),be=Ce.default`
  mutation addBundleProductsToCart($input: AddBundleProductsToCartInput) {
    addBundleProductsToCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;function a(e){let{getCustomerToken:r,getStore:o,getCurrency:t}=e.config.state;return n(n(n({},r()&&{Authorization:`Bearer ${r()}`}),o()&&{store:o()}),t()&&{"Content-Currency":t()})}var he=async(e,r,o={addBundleProductsToCart:"addBundleProductsToCart"})=>{let{addBundleProductsToCart:t}=e.extendQuery(o,{addBundleProductsToCart:{query:be,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var ve=s(require("graphql-tag")),Pe=ve.default`
  mutation addConfigurableProductsToCart($input: AddConfigurableProductsToCartInput) {
    addConfigurableProductsToCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function Q(e,r,o={addConfigurableProductsToCart:"addConfigurableProductsToCart"}){let{addConfigurableProductsToCart:t}=e.extendQuery(o,{addConfigurableProductsToCart:{query:Pe,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var xe=s(require("graphql-tag")),Qe=xe.default`
  mutation addProductsToCart($cartId: String!, $cartItems: [CartItemInput!]!) {
    addProductsToCart(cartId,: $cartId, cartItems,: $cartItems) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
      user_errors {
        code
        message
      }
    }
  }
`;async function w(e,r,o={addProductsToCart:"addProductsToCart"}){let{addProductsToCart:t}=e.extendQuery(o,{addProductsToCart:{query:Qe,variables:n({},r)}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var we=s(require("graphql-tag")),qe=we.default`
  mutation addProductsToWishlist($id: ID!, $items: [WishlistItemInput!]!) {
    addProductsToWishlist(wishlistId: $id, wishlistItems: $items) {
      wishlist {
        id
        items_count
        sharing_code
        items_v2 {
          items {
            id
            quantity
            description
            added_at
            product {
              ...on ConfigurableProduct {
                configurable_options {
                  attribute_code
                  attribute_uid
                  label
                  position
                  uid
                  use_default
                  values {
                    label
                    swatch_data {
                      value
                    }
                    uid
                  }
                }
              }
              ... on BundleProduct {
                items {
                  sku
                  title
                  options {
                    uid
                    quantity
                    product {
                      uid
                      sku
                      name
                      price_range {
                        maximum_price {
                          final_price {
                            currency
                            value
                          }
                          regular_price {
                            currency
                            value
                          }
                        }
                        minimum_price {
                          final_price {
                            currency
                            value
                          }
                          regular_price {
                            currency
                            value
                          }
                        }
                      }
                    }
                  }
                }
              }
              uid
              __typename
              sku
              name
              stock_status
              only_x_left_in_stock
              rating_summary
              categories {
                uid
                name
                url_suffix
                url_path
                breadcrumbs {
                  category_name,
                  category_url_path
                }
              }
              price_range {
                maximum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
                minimum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
              }
              thumbnail {
                url
                position
                disabled
                label
              }
              url_key
              url_rewrites {
                url
              }
              review_count
              reviews {
                items {
                  average_rating
                  ratings_breakdown {
                    name
                    value
                  }
                }
              }
              ... on GroupedProduct {
                items {
                  product {
                    sku
                  }
                }
              }
            }
          }
          page_info {
            current_page
            page_size
            total_pages
          }
        }
      }
    }
  }
`;var Ie=async(e,r,o={addProductsToWishlist:"addProductsToWishlist"})=>{let{addProductsToWishlist:t}=e.extendQuery(o,{addProductsToWishlist:{query:qe,variables:n({},r)}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var Ae=s(require("graphql-tag")),ke=Ae.default`
  mutation addSimpleProductsToCart($input: AddSimpleProductsToCartInput) {
    addSimpleProductsToCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;var Me=async(e,r,o={addSimpleProductsToCart:"addSimpleProductsToCart"})=>{let{addSimpleProductsToCart:t}=e.extendQuery(o,{addSimpleProductsToCart:{query:ke,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var Re=s(require("graphql-tag")),Se=Re.default`
  mutation addDownloadableProductsToCart($input: AddDownloadableProductsToCartInput) {
    addDownloadableProductsToCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function q(e,r,o={addDownloadableProductsToCart:"addDownloadableProductsToCart"}){let{addDownloadableProductsToCart:t}=e.extendQuery(o,{addDownloadableProductsToCart:{query:Se,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Te=s(require("graphql-tag")),$e=Te.default`
  mutation addVirtualProductsToCart($input: AddVirtualProductsToCartInput) {
    addVirtualProductsToCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function I(e,r,o={addVirtualProductsToCart:"addVirtualProductsToCart"}){let{addVirtualProductsToCart:t}=e.extendQuery(o,{addVirtualProductsToCart:{query:$e,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Le=s(require("graphql-tag")),Ee=Le.default`
  mutation applyCouponToCart($input: ApplyCouponToCartInput) {
    applyCouponToCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function A(e,r,o={applyCouponToCart:"applyCouponToCart"}){let{applyCouponToCart:t}=e.extendQuery(o,{applyCouponToCart:{query:Ee,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Oe=s(require("graphql-tag")),Fe=Oe.default`
  query availableStores {
    availableStores {
      absolute_footer
      allow_guests_to_write_product_reviews
      allow_items
      allow_order
      autocomplete_on_storefront
      base_currency_code
      base_link_url
      base_media_url
      base_static_url
      base_url
      catalog_default_sort_by
      category_fixed_product_tax_display_setting
      category_url_suffix
      cms_home_page
      cms_no_cookies
      cms_no_route
      code
      configurable_thumbnail_source
      copyright
      default_description
      default_display_currency_code
      default_keywords
      default_title
      front
      grid_per_page
      grid_per_page_values
      head_includes
      head_shortcut_icon
      header_logo_src
      id
      is_default_store
      is_default_store_group
      list_mode
      list_per_page
      list_per_page_values
      locale
      logo_alt
      logo_height
      logo_width
      magento_wishlist_general_is_enabled
      minimum_password_length
      no_route
      payment_payflowpro_cc_vault_active
      product_fixed_product_tax_display_setting
      product_reviews_enabled
      product_url_suffix
      required_character_classes_number
      root_category_id
      root_category_uid
      sales_fixed_product_tax_display_setting
      secure_base_link_url
      secure_base_media_url
      secure_base_static_url
      secure_base_url
      show_cms_breadcrumbs
      store_code
      store_group_code
      store_group_name
      store_name
      store_sort_order
      timezone
      title_prefix
      title_separator
      title_suffix
      use_store_in_url
      website_code
      website_id
      website_name
      weight_unit
      welcome
    }
  }
`;var Ve=async(e,r={availableStores:"availableStores"})=>{let{availableStores:o}=e.extendQuery(r,{availableStores:{query:Fe}});return e.client.query({query:o.query,context:{headers:a(e)}})};var Ge=s(require("graphql-tag")),Be=Ge.default`
  query cart($cartId: String!) {
    cart(cart_id:$cartId) {
      id
      email
      is_virtual
      applied_coupons {
        code
      }
      prices {
        subtotal_excluding_tax {
          value
        },
        subtotal_including_tax {
          value
        },
        applied_taxes {
          amount {
            value
          },
          label
        }
        discounts {
          amount {
            value
          },
          label
        }
        grand_total {
          value
        }
      }
      items {
        uid
        product {
          uid
          __typename
          sku
          name
          stock_status
          only_x_left_in_stock
          rating_summary
          thumbnail {
            url
            position
            disabled
            label
          }
          url_key
          url_rewrites {
            url
          }
          price_range {
            maximum_price {
              final_price {
                currency
                value
              }
              regular_price {
                currency
                value
              }
            }
            minimum_price {
              final_price {
                currency
                value
              }
              regular_price {
                currency
                value
              }
            }
          }
          categories {
            uid
            name
            url_suffix
            url_path
            breadcrumbs {
              category_name,
              category_url_path
            }
          }
          review_count
          reviews {
            items {
              average_rating
              ratings_breakdown {
                name
                value
              }
            }
          }
        }
        prices {
          row_total {
            value
          }
          row_total_including_tax {
            value
          }
          total_item_discount {
            value
          }
        }
        quantity
        ... on ConfigurableCartItem {
          configurable_options {
            configurable_product_option_uid
            option_label
            configurable_product_option_value_uid
            value_label
          }
          configured_variant {
            sku
            thumbnail {
              url
            }
          }
        }
        ... on BundleCartItem {
          bundle_options {
            uid
            label
            type
            values {
              id
              label
              price
              quantity
            }
          }
        }
      }
      total_quantity
      shipping_addresses {
        firstname
        lastname
        street
        city
        company
        region {
          code
          region_id
          label
        }
        postcode
        telephone
        country {
          code
          label
        }
        selected_shipping_method {
          carrier_code
          carrier_title
          method_code
          method_title
          amount {
            value
            currency
          }
        }
      }
      billing_address {
        firstname
        lastname
        street
        city
        company
        region {
          code
          region_id
          label
        }
        postcode
        telephone
        country {
          code
          label
        }
      }
    }
  }
`;async function k(e,r,o={cart:"cart"}){let{cart:t}=e.extendQuery(o,{cart:{query:Be,variables:{cartId:r??""}}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var ze=s(require("graphql-tag")),He=ze.default`
  query cartTotalQty($cartId: String!) {
    cart(cart_id:$cartId) {
      total_quantity
    }
  }
`;var De=async(e,r,o={cartTotalQty:"cartTotalQty"})=>{let{cartTotalQty:t}=e.extendQuery(o,{cartTotalQty:{query:He,variables:{cartId:r??""}}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})};var Ue=s(require("graphql-tag")),Ne=Ue.default`
  query categoryList {
    categories {
      items {
        children {
          include_in_menu
          is_anchor
          level
          name
          position
          product_count
          uid
          url_path
          url_suffix
          children {
            include_in_menu
            is_anchor
            level
            name
            position
            product_count
            uid
            url_path
            url_suffix
            children {
              include_in_menu
              is_anchor
              level
              name
              position
              product_count
              uid
              url_path
              url_suffix
            }
          }
        }
        product_count
        name
        uid
      }
    }
  }
`;async function M(e,r,o={categoryList:"categoryList"}){let{categoryList:t}=e.extendQuery(o,{categoryList:{query:Ne,variables:n({},r)}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var We=s(require("graphql-tag")),Ye=We.default`
  query categorySearch($filters: CategoryFilterInput) {
    categoryList(filters: $filters) {
      include_in_menu
      is_anchor
      level
      name
      position
      product_count
      uid
      url_path
      url_suffix
    }
  }
`;async function R(e,r,o={categorySearch:"categorySearch"}){let{categorySearch:t}=e.extendQuery(o,{categorySearch:{query:Ye,variables:n({},r)}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var Je=s(require("graphql-tag")),je=Je.default`
  mutation changeCustomerPassword($currentPassword: String!, $newPassword: String!) {
    changeCustomerPassword(
      currentPassword: $currentPassword
      newPassword: $newPassword
    ) {
      date_of_birth
      default_billing
      default_shipping
      email
      firstname
      is_subscribed
      lastname
      middlename
      prefix
      suffix
      taxvat
      addresses {
        city
        country_code
        default_billing
        default_shipping
        extension_attributes {
          attribute_code
          value
        }
        firstname
        id
        lastname
        postcode
        prefix
        region {
          region_code
          region_id
          region
        }
        street
        suffix
        telephone
        vat_id
      }
    }
  }
`;var Ke=async(e,r,o={changeCustomerPassword:"changeCustomerPassword"})=>{var t;try{let{changeCustomerPassword:i}=e.extendQuery(o,{changeCustomerPassword:{query:je,variables:n({},r)}});return await e.client.mutate({mutation:i.query,variables:i.variables,context:{headers:a(e)}})}catch(i){if(i.graphQLErrors)return{errors:i.graphQLErrors,data:null};throw((t=i.networkError)==null?void 0:t.result)||i}};var Xe=s(require("graphql-tag")),Ze=Xe.default`
  query cmsBlock($identifiers: [String]) {
      cmsBlocks(identifiers: $identifiers) {
          items {
              content
              identifier
              title
          }
      }
  }
`;async function S(e,r,o={cmsBlocks:"cmsBlocks"}){let{cmsBlocks:t}=e.extendQuery(o,{cmsBlocks:{query:Ze,variables:{identifiers:r}}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var T=require("@vue-storefront/core");var et=s(require("graphql-tag")),tt=et.default`
  query cmsPage($identifier: String) {
    cmsPage(identifier:$identifier) {
      identifier
      content
      title
      meta_title
      meta_description
      meta_keywords
      content_heading
    }
  }
`;async function $(e,r,o={cmsPage:"cmsPage"}){var t;try{let{cmsPage:i}=e.extendQuery(o,{cmsPage:{query:tt,variables:{identifier:r}}});return await e.client.query({query:i.query,variables:i.variables,context:{headers:a(e)}})}catch(i){if(i.graphQLErrors)return T.Logger.debug(i),y(n({},i),{errors:i.graphQLErrors,data:null});throw T.Logger.error(i),((t=i.networkError)==null?void 0:t.result)||i}}var rt=s(require("graphql-tag")),ot=rt.default`
  query countriesList {
    countries {
      id
      two_letter_abbreviation
      full_name_locale
      full_name_english
    }
  }
`;async function L(e,r={countries:"countries"}){let{countries:o}=e.extendQuery(r,{countries:{query:ot}});return e.client.query({query:o.query,context:{headers:a(e)}})}var at=s(require("graphql-tag")),it=at.default`
  query countryInformation($id: String) {
    country(id: $id) {
      id
      two_letter_abbreviation
      full_name_locale
      full_name_english
      available_regions {
        id
        code
        name
      }
    }
  }
`;async function E(e,r,o={country:"country"}){let{country:t}=e.extendQuery(o,{country:{query:it,variables:{id:r}}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var nt=require("graphql");var _=async(e,r)=>{try{let{secretkey:o}=e.config.recaptcha,t=`https://www.google.com/recaptcha/api/siteverify?secret=${o}&response=${r}`;return await(await fetch(t)).json()}catch(o){throw o.message||o}};var ut=s(require("graphql-tag")),st=ut.default`
  mutation createCustomer($input: CustomerCreateInput!) {
    createCustomerV2(input: $input) {
      customer {
        date_of_birth
        default_billing
        default_shipping
        email
        firstname
        is_subscribed
        lastname
        middlename
        prefix
        suffix
        taxvat
        addresses {
          city
          country_code
          default_billing
          default_shipping
          extension_attributes {
            attribute_code
            value
          }
          firstname
          id
          lastname
          postcode
          prefix
          region {
            region_code
            region_id
            region
          }
          street
          suffix
          telephone
          vat_id
        }
      }
    }
  }
`;var lt=async(e,r,o={createCustomer:"createCustomer"})=>{var i;try{let t=r,{recaptchaToken:u}=t,l=f(t,["recaptchaToken"]);if(e.config.recaptcha.isEnabled&&!(await _(e,u)).success)return{errors:[new nt.GraphQLError("Error during reCaptcha verification. Please try again.")],data:null};let{createCustomer:m}=e.extendQuery(o,{createCustomer:{query:st,variables:{input:l}}});return await e.client.mutate({mutation:m.query,variables:m.variables,context:{headers:a(e)}})}catch(u){if(u.graphQLErrors)return{errors:u.graphQLErrors,data:null};throw((i=u.networkError)==null?void 0:i.result)||u}};var mt=s(require("graphql-tag")),dt=mt.default`
  mutation createCustomerAddress($input: CustomerAddressInput!) {
    createCustomerAddress(input: $input) {
      id
      region {
        region
        region_id
        region_code
      }
      country_code
      street
      telephone
      postcode
      city
      default_shipping
      default_billing
      vat_id
    }
  }
`;async function O(e,r,o={createCustomerAddress:"createCustomerAddress"}){let{createCustomerAddress:t}=e.extendQuery(o,{createCustomerAddress:{query:dt,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var ct=s(require("graphql-tag")),pt=ct.default`
  mutation createEmptyCart {
    createEmptyCart
  }
`;var _t=async(e,r={createEmptyCart:"createEmptyCart"})=>{let{createEmptyCart:o}=e.extendQuery(r,{createEmptyCart:{query:pt}});return e.client.mutate({mutation:o.query,context:{headers:a(e)}})};var gt=require("graphql");var yt=s(require("graphql-tag")),ft=yt.default`
  mutation createProductReview($input: CreateProductReviewInput!) {
    createProductReview(input: $input){
      review {
        average_rating
        ratings_breakdown {
          name
          value
        }
        nickname
        summary
        text
        created_at
      }
    }
  }
`;var Ct=async(e,r,o={createProductReview:"createProductReview"})=>{let l=r,{recaptchaToken:t}=l,i=f(l,["recaptchaToken"]);if(e.config.recaptcha.isEnabled&&!(await _(e,t)).success)return{errors:[new gt.GraphQLError("Error during reCaptcha verification. Please try again.")],data:null};let{createProductReview:u}=e.extendQuery(o,{createProductReview:{query:ft,variables:{input:i}}});return e.client.mutate({mutation:u.query,variables:u.variables,context:{headers:a(e)}})};var bt=s(require("graphql-tag")),ht=bt.default`
  query currency {
    currency{
      available_currency_codes
      base_currency_code
      base_currency_symbol
      default_display_currecy_code
      default_display_currecy_symbol
      default_display_currency_code
      default_display_currency_symbol
      exchange_rates {
        currency_to
        rate
      }
    }
  }
`;async function F(e,r={currency:"currency"}){let{currency:o}=e.extendQuery(r,{currency:{query:ht}});return e.client.query({query:o.query,context:{headers:a(e)}})}var vt=s(require("graphql-tag")),Pt=vt.default`
  query customer {
    customer {
      date_of_birth
      default_billing
      default_shipping
      email
      firstname
      is_subscribed
      lastname
      middlename
      prefix
      suffix
      taxvat
      addresses {
        city
        country_code
        default_billing
        default_shipping
        extension_attributes {
          attribute_code
          value
        }
        firstname
        id
        lastname
        postcode
        prefix
        region {
          region_code
          region_id
          region
        }
        street
        suffix
        telephone
        vat_id
      }
      is_subscribed
    }
  }
`;var xt=async(e,r={customer:"customer"})=>{let{customer:o}=e.extendQuery(r,{customer:{query:Pt}});return e.client.query({query:o.query,context:{headers:a(e)}})};var Qt=s(require("graphql-tag")),wt=Qt.default`
  query customerCart {
    customerCart {
      id
      email
      is_virtual
      applied_coupons {
        code
      }
      prices {
        subtotal_excluding_tax {
          value
        },
        subtotal_including_tax {
          value
        },
        applied_taxes {
          amount {
            value
          },
          label
        }
        discounts {
          amount {
            value
          },
          label
        }
        grand_total {
          value
        }
      }
      items {
        uid
        product {
          uid
          __typename
          sku
          name
          stock_status
          only_x_left_in_stock
          rating_summary
          thumbnail {
            url
            position
            disabled
            label
          }
          url_key
          url_rewrites {
            url
          }
          price_range {
            maximum_price {
              final_price {
                currency
                value
              }
              regular_price {
                currency
                value
              }
            }
            minimum_price {
              final_price {
                currency
                value
              }
              regular_price {
                currency
                value
              }
            }
          }
          categories {
            uid
            name
            url_suffix
            url_path
            breadcrumbs {
              category_name,
              category_url_path
            }
          }
          review_count
          reviews {
            items {
              average_rating
              ratings_breakdown {
                name
                value
              }
            }
          }
        }
        prices {
          row_total {
            value
          }
          row_total_including_tax {
            value
          }
          total_item_discount {
            value
          }
        }
        quantity
        ... on ConfigurableCartItem {
          configurable_options {
            configurable_product_option_uid
            option_label
            configurable_product_option_value_uid
            value_label
          }
          configured_variant {
            sku
            thumbnail {
              url
            }
          }
        }
        ... on BundleCartItem {
          bundle_options {
            uid
            label
            type
            values {
              id
              label
              price
              quantity
            }
          }
        }
      }
      total_quantity
      shipping_addresses {
        firstname
        lastname
        street
        city
        company
        region {
          code
          region_id
          label
        }
        postcode
        telephone
        country {
          code
          label
        }
        selected_shipping_method {
          carrier_code
          carrier_title
          method_code
          method_title
          amount {
            value
            currency
          }
        }
      }
      billing_address {
        firstname
        lastname
        street
        city
        company
        region {
          code
          region_id
          label
        }
        postcode
        telephone
        country {
          code
          label
        }
      }
    }
  }
`;async function V(e,r={customerCart:"customerCart"}){let{customerCart:o}=e.extendQuery(r,{customerCart:{query:wt}});return e.client.query({query:o.query,context:{headers:a(e)}})}var qt=s(require("graphql-tag")),It=qt.default`
  query customerOrders($currentPage: Int = 1, $filter: CustomerOrdersFilterInput = {}, $pageSize: Int = 10) {
    customer {
      orders(currentPage: $currentPage, filter: $filter, pageSize: $pageSize) {
        items {
          number
          id
          order_date
          grand_total
          total {
            discounts {
              amount {
                currency
                value
              }
              label
            }
            base_grand_total {
              currency
              value
            }
            shipping_handling {
              amount_excluding_tax {
                currency
                value
              }
              amount_including_tax {
                currency
                value
              }
              discounts {
                amount {
                  currency
                  value
                }
              }
              taxes {
                amount{
                  currency
                  value
                }
                rate
                title
              }
              total_amount {
                currency
                value
              }
            }
            subtotal {
              currency
              value
            }
            taxes {
              amount {
                currency
                value
              }
              rate
              title
            }
            total_shipping {
              currency
              value
            }
            total_tax {
              currency
              value
            }
          }
          status
          comments {
            message
            timestamp
          }
          invoices {
            comments {
              message
              timestamp
            }
            id
            items {
              discounts {
                amount {
                  currency
                  value
                }
                label
              }
              id
              product_name
              product_sale_price {
                currency
                value
              }
              product_sku
              quantity_invoiced
            }
            number
            total {
              discounts {
                amount {
                  currency
                  value
                }
                label
              }
              base_grand_total {
                currency
                value
              }
              shipping_handling {
                amount_excluding_tax {
                  currency
                  value
                }
                amount_including_tax {
                  currency
                  value
                }
                discounts {
                  amount {
                    currency
                    value
                  }
                }
                taxes {
                  amount{
                    currency
                    value
                  }
                  rate
                  title
                }
                total_amount {
                  currency
                  value
                }
              }
              subtotal {
                currency
                value
              }
              taxes {
                amount {
                  currency
                  value
                }
                rate
                title
              }
              total_shipping {
                currency
                value
              }
              total_tax {
                currency
                value
              }
            }
          }
          items {
            discounts {
              amount {
                currency
                value
              }
              label
            }
            entered_options {
              label
              value
            }
            id
            product_name
            product_sale_price {
              currency
              value
            }
            product_sku
            product_type
            product_url_key
            quantity_canceled
            quantity_invoiced
            quantity_ordered
            quantity_refunded
            quantity_returned
            quantity_shipped
            selected_options {
              label
              value
            }
            status
          }
          payment_methods {
            name
            type
            additional_data {
              name
              value
            }
          }
          shipments {
            comments {
              message
              timestamp
            }
            id
            number
            tracking {
              carrier
              number
              title
            }
            items {
              id
              product_name
              product_sale_price {
                currency
                value
              }
              product_sku
              quantity_shipped
            }
          }
          shipping_address {
            city
            country_code
            firstname
            lastname
            postcode
            prefix
            region
            street
            suffix
            telephone
          }
          billing_address {
            city
            country_code
            firstname
            lastname
            postcode
            prefix
            region
            street
            suffix
            telephone
          }
          shipping_method
        }
        page_info {
            current_page
            total_pages
            page_size
        }
        total_count
      }
    }
  }
`;var At=async(e,r,o={customerOrders:"customerOrders"})=>{var l,m;let t=n({pageSize:10,currentPage:1},r),i={pageSize:t.pageSize<=0?10:t.pageSize,currentPage:t.currentPage<=0?1:t.currentPage};t.filter&&(i.filter=t.filter);let{customerOrders:u}=e.extendQuery(o,{customerOrders:{query:It,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}};var kt=s(require("graphql-tag")),Mt=kt.default`
  query customerProductReview($pageSize: Int = 10, $currentPage: Int = 1) {
    customer {
        reviews(pageSize: $pageSize, currentPage: $currentPage) {
          items {
            average_rating
            ratings_breakdown {
              name
              value
            }
            nickname
            summary
            text
            created_at
            product {
              name
              uid
            }
          }
          page_info {
            current_page
            page_size
            total_pages
          }
        }
    }
  }
`;var Rt=async(e,r,o={reviews:"reviews"})=>{var l,m;let t={pageSize:10,currentPage:1},i={pageSize:t.pageSize,currentPage:t.currentPage},{reviews:u}=e.extendQuery(o,{reviews:{query:Mt,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}};var St=s(require("graphql-tag"));var Tt=async(e,{mutation:r,mutationVariables:o,fetchPolicy:t})=>e.client.mutate({mutation:St.default`${r}`,variables:n({},o),fetchPolicy:t||"no-cache",context:{headers:a(e)}});var $t=s(require("graphql-tag"));var Lt=async(e,{query:r,queryVariables:o,fetchPolicy:t})=>e.client.query({query:$t.default`${r}`,variables:n({},o),fetchPolicy:t||"no-cache",context:{headers:a(e)}});var Et=s(require("graphql-tag")),Ot=Et.default`
  mutation deleteCustomerAddress($id: Int!) {
    deleteCustomerAddress(id: $id)
  }
`;async function G(e,r,o={deleteCustomerAddress:"deleteCustomerAddress"}){let{deleteCustomerAddress:t}=e.extendQuery(o,{deleteCustomerAddress:{query:Ot,variables:{id:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Gt=require("graphql");var Ft=s(require("graphql-tag")),Vt=Ft.default`
  mutation generateCustomerToken($email: String!, $password: String!) {
    generateCustomerToken(email: $email, password: $password) {
      token
    }
  }
`;var Bt=async(e,r,o={generateCustomerToken:"generateCustomerToken"})=>{var t;try{if(e.config.recaptcha.isEnabled&&!(await _(e,r.recaptchaToken)).success)return{errors:[new Gt.GraphQLError("Error during reCaptcha verification. Please try again.")],data:null};let{generateCustomerToken:i}=e.extendQuery(o,{generateCustomerToken:{query:Vt,variables:{email:r.email,password:r.password}}});return await e.client.mutate({mutation:i.query,variables:i.variables,context:{headers:a(e)}})}catch(i){if(i.graphQLErrors)return{errors:i.graphQLErrors,data:null};throw((t=i.networkError)==null?void 0:t.result)||i}};var zt=s(require("graphql-tag")),Ht=zt.default`
  query CustomerAvailablePaymentMethods {
    customerCart {
      available_payment_methods {
        code
        title
      }
    }
  }
`;var Dt=async(e,r,o={paymentMethods:"paymentMethods"})=>{var i,u;let{paymentMethods:t}=e.extendQuery(o,{paymentMethods:{query:Ht}});try{return await e.client.query({query:t.query,context:{headers:a(e)}})}catch(l){throw((i=l.graphQLErrors)==null?void 0:i[0].message)||((u=l.networkError)==null?void 0:u.result)||l}};var Ut=s(require("graphql-tag")),Nt=Ut.default`
  query CustomerAvailableShippingMethods {
    customerCart {
      shipping_addresses {
        available_shipping_methods {
          amount {
            currency
            value
          }
          available
          carrier_code
          carrier_title
          error_message
          method_code
          method_title
          price_excl_tax {
            currency
            value
          }
          price_incl_tax {
            currency
            value
          }
        }
      }
    }
  }
`;async function B(e,r={shippingMethods:"shippingMethods"}){var t,i;let{shippingMethods:o}=e.extendQuery(r,{shippingMethods:{query:Nt}});try{return await e.client.query({query:o.query,context:{headers:a(e)}})}catch(u){throw((t=u.graphQLErrors)==null?void 0:t[0].message)||((i=u.networkError)==null?void 0:i.result)||u}}var Wt=s(require("graphql-tag")),Yt=Wt.default`
  query GuestAvailablePaymentMethods($cartId: String!) {
    cart(cart_id: $cartId) {
      available_payment_methods {
        code
        title
      }
    }
  }
`;async function z(e,r,o={paymentMethods:"paymentMethods"}){var u,l;let t={cartId:r.cartId||""},{paymentMethods:i}=e.extendQuery(o,{paymentMethods:{query:Yt,variables:t}});try{return await e.client.query({query:i.query,variables:i.variables,context:{headers:a(e)}})}catch(m){throw((u=m.graphQLErrors)==null?void 0:u[0].message)||((l=m.networkError)==null?void 0:l.result)||m}}var Jt=s(require("graphql-tag")),jt=Jt.default`
  query GuestAvailableShippingMethods($cartId: String!) {
    cart(cart_id:$cartId) {
      shipping_addresses {
        available_shipping_methods {
          amount {
            currency
            value
          }
          available
          carrier_code
          carrier_title
          error_message
          method_code
          method_title
          price_excl_tax {
            currency
            value
          }
          price_incl_tax {
            currency
            value
          }
        }
      }
    }
  }
`;var Kt=async(e,r,o={shippingMethods:"shippingMethods"})=>{var i,u;let{shippingMethods:t}=e.extendQuery(o,{shippingMethods:{query:jt,variables:n({},r)}});try{return await e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}catch(l){throw((i=l.graphQLErrors)==null?void 0:i[0].message)||((u=l.networkError)==null?void 0:u.result)||l}};var Xt=s(require("graphql-tag")),Zt=Xt.default`
  query getCustomerAddresses {
    customer {
      addresses {
        city
        country_code
        default_billing
        default_shipping
        extension_attributes {
          attribute_code
          value
        }
        firstname
        id
        lastname
        postcode
        prefix
        region {
          region_code
          region_id
          region
        }
        street
        suffix
        telephone
        vat_id
      }
    }
  }
`;async function H(e,r={getCustomerAddresses:"getCustomerAddresses"}){var t,i;let{getCustomerAddresses:o}=e.extendQuery(r,{getCustomerAddresses:{query:Zt}});try{return await e.client.query({query:o.query,context:{headers:a(e)}})}catch(u){throw((t=u.graphQLErrors)==null?void 0:t[0].message)||((i=u.networkError)==null?void 0:i.result)||u}}var er=s(require("graphql-tag")),tr=er.default`
  mutation mergeCarts($sourceCartId: String!, $destinationCartId: String!) {
    mergeCarts(source_cart_id: $sourceCartId, destination_cart_id: $destinationCartId) {
      id
      email
      is_virtual
      applied_coupons {
        code
      }
      prices {
        subtotal_excluding_tax {
          value
        },
        subtotal_including_tax {
          value
        },
        applied_taxes {
          amount {
            value
          },
          label
        }
        discounts {
          amount {
            value
          },
          label
        }
        grand_total {
          value
        }
      }
      items {
        uid
        product {
          uid
          __typename
          sku
          name
          stock_status
          only_x_left_in_stock
          rating_summary
          thumbnail {
            url
            position
            disabled
            label
          }
          url_key
          url_rewrites {
            url
          }
          price_range {
            maximum_price {
              final_price {
                currency
                value
              }
              regular_price {
                currency
                value
              }
            }
            minimum_price {
              final_price {
                currency
                value
              }
              regular_price {
                currency
                value
              }
            }
          }
          categories {
            uid
            name
            url_suffix
            url_path
            breadcrumbs {
              category_name,
              category_url_path
            }
          }
          review_count
          reviews {
            items {
              average_rating
              ratings_breakdown {
                name
                value
              }
            }
          }
        }
        prices {
          row_total {
            value
          }
          row_total_including_tax {
            value
          }
          total_item_discount {
            value
          }
        }
        quantity
        ... on ConfigurableCartItem {
          configurable_options {
            configurable_product_option_uid
            option_label
            configurable_product_option_value_uid
            value_label
          }
          configured_variant {
            sku
            thumbnail {
              url
            }
          }
        }
        ... on BundleCartItem {
          bundle_options {
            uid
            label
            type
            values {
              id
              label
              price
              quantity
            }
          }
        }
      }
      total_quantity
      shipping_addresses {
        firstname
        lastname
        street
        city
        company
        region {
          code
          region_id
          label
        }
        postcode
        telephone
        country {
          code
          label
        }
        selected_shipping_method {
          carrier_code
          carrier_title
          method_code
          method_title
          amount {
            value
            currency
          }
        }
      }
      billing_address {
        firstname
        lastname
        street
        city
        company
        region {
          code
          region_id
          label
        }
        postcode
        telephone
        country {
          code
          label
        }
      }
    }
  }
`;var rr=async(e,r,o={mergeCarts:"mergeCarts"})=>{let{mergeCarts:t}=e.extendQuery(o,{mergeCarts:{query:tr,variables:{sourceCartId:r.sourceCartId,destinationCartId:r.destinationCartId}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var or=s(require("graphql-tag")),ar=or.default`
  mutation placeOrder($input: PlaceOrderInput) {
    placeOrder(input: $input) {
      order {
        order_number
      }
    }
  }
`;async function D(e,r,o={placeOrder:"placeOrder"}){var i,u;let{placeOrder:t}=e.extendQuery(o,{placeOrder:{query:ar,variables:{input:r}}});try{return await e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}catch(l){throw((i=l.graphQLErrors)==null?void 0:i[0].message)||((u=l.networkError)==null?void 0:u.result)||l}}var U=require("@vue-storefront/core");var ir=s(require("graphql-tag")),ur=ir.default`
  query productDetails(
    $search: String = "",
    $filter: ProductAttributeFilterInput,
    $pageSize: Int = 10,
    $currentPage: Int = 1,
    $sort: ProductAttributeSortInput
  ) {
    products(search: $search, filter: $filter, sort: $sort, pageSize: $pageSize, currentPage: $currentPage) {
      items {
        uid
        sku
        name
        stock_status
        only_x_left_in_stock
        thumbnail {
          url
          position
          disabled
          label
        }
        url_key
        url_rewrites {
          url
        }
        categories {
          uid
          name
          url_suffix
          url_path
          breadcrumbs {
            category_name,
            category_url_path
          }
        }
        rating_summary
        review_count
        reviews {
          items {
            average_rating
            ratings_breakdown {
              name
              value
            }
          }
        }
        small_image {
          url
          position
          disabled
          label
        }
        image {
          url
          position
          disabled
          label
        }
        media_gallery {
          url
          position
          disabled
          label
        }
        thumbnail {
          url
          position
          disabled
          label
        }
        url_key
        url_rewrites {
          url
        }
        meta_description
        meta_keyword
        meta_title
        description {
          html
        }
        short_description {
          html
        }
        options_container
        special_to_date
        ... on ConfigurableProduct {
          configurable_options {
            attribute_code
            attribute_uid
            label
            position
            uid
            use_default
            values {
              label
              swatch_data {
                value
              }
              uid
            }
          }
        }

        ... on DownloadableProduct {
          downloadable_product_samples {
            sample_url
            title
          }
          downloadable_product_links {
            id
            price
            title
            uid
          }
        }
        ... on VirtualProduct {
          gift_message_available
          product_links {
            link_type
            linked_product_sku
            linked_product_type
            position
            sku
          }
        }
        ... on GroupedProduct {
          items {
            position
            qty
            product {
              uid
              sku
              name
              stock_status
              only_x_left_in_stock
              price_range {
                maximum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
                minimum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
              }
              thumbnail {
                url
                position
                disabled
                label
              }
            }
          }
        }
      }
    }
  }
`;async function N(e,r,o={productDetail:"productDetail"}){var l;let t=n({pageSize:10,currentPage:1},r),i=y(n({},t),{pageSize:t.pageSize,currentPage:t.currentPage});t.search&&(i.search=t.search),t.filter&&(i.filter=t.filter),t.sort&&(i.sort=t.sort);let{productDetail:u}=e.extendQuery(o,{productDetail:{query:ur,variables:i}});try{let m=await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}});if(m.data.products.items.length===0)throw new Error("No products found");return m}catch(m){if(m.graphQLErrors)return U.Logger.debug(m),y(n({},m),{errors:m.graphQLErrors,data:null});throw U.Logger.error(m),((l=m.networkError)==null?void 0:l.result)||m}}var sr=s(require("graphql-tag")),nr=sr.default`
  query productReview($search: String = "", $filter: ProductAttributeFilterInput, $pageSize: Int = 10, $currentPage: Int = 1, $sort: ProductAttributeSortInput) {
    products(search: $search, filter: $filter, sort: $sort) {
      items {
        review_count
        reviews(pageSize: $pageSize, currentPage: $currentPage) {
          items {
            average_rating
            ratings_breakdown {
              name
              value
            }
            product {
              name
              uid
            }
            nickname
            summary
            text
            created_at
          }
          page_info {
            current_page
            page_size
            total_pages
          }
        }
      }
    }
  }
`;var lr=async(e,r,o={productReview:"productReview"})=>{var l,m;let t=n({pageSize:10,currentPage:1},r),i={pageSize:t.pageSize,currentPage:t.currentPage};t.search&&(i.search=t.search),t.filter&&(i.filter=t.filter),t.sort&&(i.sort=t.sort);let{productReview:u}=e.extendQuery(o,{productReview:{query:nr,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}};var mr=s(require("graphql-tag")),dr=mr.default`
  query productReviewRatingsMetadata{
    productReviewRatingsMetadata {
      items {
        id
        name
        values {
          value_id
          value
        }
      }
    }
  }
`;var cr=async(e,r={productReviewRatingsMetadata:"productReviewRatingsMetadata"})=>{let{productReviewRatingsMetadata:o}=e.extendQuery(r,{productReviewRatingsMetadata:{query:dr}});return e.client.query({query:o.query,context:{headers:a(e)}})};var pr=s(require("graphql-tag")),_r=pr.default`
  query productsList($search: String = "", $filter: ProductAttributeFilterInput, $pageSize: Int = 10, $currentPage: Int = 1, $sort: ProductAttributeSortInput) {
    products(search: $search, filter: $filter, pageSize: $pageSize, currentPage: $currentPage, sort: $sort) {
      aggregations {
        attribute_code
        label
        options {
          label
          value
          count
        }
      }
      items {
        uid
        sku
        name
        stock_status
        only_x_left_in_stock
        rating_summary
        thumbnail {
          url
          position
          disabled
          label
        }
        url_key
        url_rewrites {
          url
        }
        price_range {
          maximum_price {
            final_price {
              currency
              value
            }
            regular_price {
              currency
              value
            }
          }
          minimum_price {
            final_price {
              currency
              value
            }
            regular_price {
              currency
              value
            }
          }
        }
        categories {
          uid
          name
          url_suffix
          url_path
          breadcrumbs {
            category_name,
            category_url_path
          }
        }
        review_count
        reviews {
          items {
            average_rating
            ratings_breakdown {
              name
              value
            }
          }
        }
        ... on GroupedProduct {
          items {
            product {
              sku
            }
          }
        }
      }
      page_info {
        current_page
        page_size
        total_pages
      }
      total_count
    }
  }
`;async function W(e,r,o={products:"products"}){var l,m;let t=n({pageSize:10,currentPage:1},r),i={pageSize:t.pageSize<=0?10:t.pageSize,currentPage:t.currentPage<=0?1:t.currentPage};t.search&&(i.search=t.search),t.filter&&(i.filter=t.filter),t.sort&&(i.sort=t.sort);let{products:u}=e.extendQuery(o,{products:{query:_r,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}}var yr=s(require("graphql-tag")),fr=yr.default`
  query relatedProduct(
    $search: String = ""
    $filter: ProductAttributeFilterInput
    $pageSize: Int = 10
    $currentPage: Int = 1
    $sort: ProductAttributeSortInput
  ) {
    products(
      search: $search
      filter: $filter
      sort: $sort
      pageSize: $pageSize
      currentPage: $currentPage
    ) {
      items {
        related_products {
          uid
    __typename
    sku
    name
    stock_status
    only_x_left_in_stock
    rating_summary
    thumbnail {
      url
      position
      disabled
      label
    }
    url_key
    url_rewrites {
      url
    }
    price_range {
      maximum_price {
        final_price {
          currency
          value
        }
        regular_price {
          currency
          value
        }
      }
      minimum_price {
        final_price {
          currency
          value
        }
        regular_price {
          currency
          value
        }
      }
    }
    categories {
      uid
      name
      url_suffix
      url_path
      breadcrumbs {
        category_name,
        category_url_path
      }
    }
    review_count
    reviews {
      items {
        average_rating
        ratings_breakdown {
          name
          value
        }
      }
    }
        }
        uid
      }
    }
  }
`;async function Y(e,r,o={relatedProduct:"relatedProduct"}){var l,m;let t=n({pageSize:10,currentPage:1},r),i={pageSize:t.pageSize,currentPage:t.currentPage};t.search&&(i.search=t.search),t.filter&&(i.filter=t.filter),t.sort&&(i.sort=t.sort);let{relatedProduct:u}=e.extendQuery(o,{relatedProduct:{query:fr,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}}var gr=s(require("graphql-tag")),Cr=gr.default`
  mutation removeCouponFromCart($input: RemoveCouponFromCartInput) {
    removeCouponFromCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function J(e,r,o={removeCouponFromCart:"removeCouponFromCart"}){let{removeCouponFromCart:t}=e.extendQuery(o,{removeCouponFromCart:{query:Cr,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var br=s(require("graphql-tag")),hr=br.default`
  mutation removeItemFromCart($input: RemoveItemFromCartInput) {
    removeItemFromCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function j(e,r,o={removeItemFromCart:"removeItemFromCart"}){let{removeItemFromCart:t}=e.extendQuery(o,{removeItemFromCart:{query:hr,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var vr=s(require("graphql-tag")),Pr=vr.default`
  mutation removeProductsFromWishlist($id: ID!, $items: [ID!]!) {
    removeProductsFromWishlist(wishlistId: $id, wishlistItemsIds: $items) {
      wishlist {
        id
        items_count
        sharing_code
        items_v2 {
          items {
            id
            quantity
            description
            added_at
            product {
              ...on ConfigurableProduct {
                configurable_options {
                  attribute_code
                  attribute_uid
                  label
                  position
                  uid
                  use_default
                  values {
                    label
                    swatch_data {
                      value
                    }
                    uid
                  }
                }
              }
              ... on BundleProduct {
                items {
                  sku
                  title
                  options {
                    uid
                    quantity
                    product {
                      uid
                      sku
                      name
                      price_range {
                        maximum_price {
                          final_price {
                            currency
                            value
                          }
                          regular_price {
                            currency
                            value
                          }
                        }
                        minimum_price {
                          final_price {
                            currency
                            value
                          }
                          regular_price {
                            currency
                            value
                          }
                        }
                      }
                    }
                  }
                }
              }
              uid
              __typename
              sku
              name
              stock_status
              only_x_left_in_stock
              rating_summary
              categories {
                uid
                name
                url_suffix
                url_path
                breadcrumbs {
                  category_name,
                  category_url_path
                }
              }
              price_range {
                maximum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
                minimum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
              }
              thumbnail {
                url
                position
                disabled
                label
              }
              url_key
              url_rewrites {
                url
              }
              review_count
              reviews {
                items {
                  average_rating
                  ratings_breakdown {
                    name
                    value
                  }
                }
              }
              ... on GroupedProduct {
                items {
                  product {
                    sku
                  }
                }
              }
            }
          }
          page_info {
            current_page
            page_size
            total_pages
          }
        }
      }
    }
  }
`;var xr=async(e,r,o={removeProductsFromWishlist:"removeProductsFromWishlist"})=>{let{removeProductsFromWishlist:t}=e.extendQuery(o,{removeProductsFromWishlist:{query:Pr,variables:n({},r)}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var qr=require("@vue-storefront/core"),Ir=require("graphql");var Qr=s(require("graphql-tag")),wr=Qr.default`
  mutation requestPasswordResetEmail($email: String!){
      requestPasswordResetEmail(email: $email)
  }
`;async function K(e,r,o={requestPasswordResetEmail:"requestPasswordResetEmail"}){let m=r,{recaptchaToken:t}=m,i=f(m,["recaptchaToken"]);if(e.config.recaptcha.isEnabled&&!(await _(e,t)).success)return{errors:[new Ir.GraphQLError("Error during reCaptcha verification. Please try again.")],data:null};let{requestPasswordResetEmail:u}=e.extendQuery(o,{requestPasswordResetEmail:{query:wr,variables:n({},i)}});qr.Logger.debug("[VSF: Magento] requestPasswordResetEmail",JSON.stringify(r,null,2));let l=await e.client.mutate({mutation:u.query,variables:u.variables,context:{headers:a(e)}});if(!l.data.requestPasswordResetEmail)throw new Error("Email was not found, or not available.");return l}var Mr=require("@vue-storefront/core"),Rr=require("graphql");var Ar=s(require("graphql-tag")),kr=Ar.default`
  mutation resetPassword($email: String!, $newPassword: String!, $resetPasswordToken: String!){
    resetPassword(email: $email, newPassword: $newPassword, resetPasswordToken: $resetPasswordToken)
  }
`;async function X(e,r,o={resetPassword:"resetPassword"}){let m=r,{recaptchaToken:t}=m,i=f(m,["recaptchaToken"]);if(e.config.recaptcha.isEnabled&&!(await _(e,t)).success)return{errors:[new Rr.GraphQLError("Error during reCaptcha verification. Please try again.")],data:null};let{resetPassword:u}=e.extendQuery(o,{resetPassword:{query:kr,variables:n({},i)}});Mr.Logger.debug("[VSF: Magento] requestPasswordResetEmail",JSON.stringify(r,null,2));let l=await e.client.mutate({mutation:u.query,variables:u.variables,context:{headers:a(e)}});if(!l.data.resetPassword)throw new Error("It was not possible to change the user password.");return l}var Sr=s(require("graphql-tag")),Tr=Sr.default`
  mutation revokeCustomerToken{
    revokeCustomerToken {
      result
    }
  }
`;var $r=async(e,r={revokeCustomerToken:"revokeCustomerToken"})=>{let{revokeCustomerToken:o}=e.extendQuery(r,{revokeCustomerToken:{query:Tr}});return e.client.mutate({mutation:o.query,context:{headers:a(e)}})};var Lr=s(require("graphql-tag")),Er=Lr.default`
  mutation setBillingAddressOnCart($input: SetBillingAddressOnCartInput) {
    setBillingAddressOnCart(input: $input) {
      cart {
        billing_address {
          firstname
          lastname
          company
          street
          city
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;var Or=async(e,r,o={setBillingAddressOnCart:"setBillingAddressOnCart"})=>{let{setBillingAddressOnCart:t}=e.extendQuery(o,{setBillingAddressOnCart:{query:Er,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var Fr=s(require("graphql-tag")),Vr=Fr.default`
  mutation setGuestEmailOnCart($input: SetGuestEmailOnCartInput) {
    setGuestEmailOnCart(input: $input) {
      cart {
        email
      }
    }
  }
`;async function Z(e,r,o={setGuestEmailOnCart:"setGuestEmailOnCart"}){let{setGuestEmailOnCart:t}=e.extendQuery(o,{setGuestEmailOnCart:{query:Vr,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Gr=s(require("graphql-tag")),Br=Gr.default`
  mutation setPaymentMethodOnCart($input: SetPaymentMethodOnCartInput) {
    setPaymentMethodOnCart(input: $input) {
      cart {
        available_payment_methods {
          code
          title
        }
        selected_payment_method {
          code
          title
        }
      }
    }
  }
`;async function ee(e,r,o={setPaymentMethodOnCart:"setPaymentMethodOnCart"}){let{setPaymentMethodOnCart:t}=e.extendQuery(o,{setPaymentMethodOnCart:{query:Br,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var zr=s(require("graphql-tag")),Hr=zr.default`
  mutation setShippingAddressesOnCart($input: SetShippingAddressesOnCartInput) {
    setShippingAddressesOnCart(input: $input) {
      cart {
        shipping_addresses {
          city
          company
          firstname
          lastname
          postcode
          region {
            code
            label
            region_id
          }
          street
          telephone
          country {
            code
            label
          }
          available_shipping_methods {
            amount {
              currency
              value
            }
            available
            carrier_code
            carrier_title
            error_message
            method_code
            method_title
            price_excl_tax {
              currency
              value
            }
            price_incl_tax {
              currency
              value
            }
          }
        }
      }
    }
  }
`;async function te(e,r,o={setShippingAddressesOnCart:"setShippingAddressesOnCart"}){let{setShippingAddressesOnCart:t}=e.extendQuery(o,{setShippingAddressesOnCart:{query:Hr,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Dr=s(require("graphql-tag")),Ur=Dr.default`
  mutation setShippingMethodsOnCart($input: SetShippingMethodsOnCartInput) {
    setShippingMethodsOnCart(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function re(e,r,o={setShippingMethodsOnCart:"setShippingMethodsOnCart"}){let{setShippingMethodsOnCart:t}=e.extendQuery(o,{setShippingMethodsOnCart:{query:Ur,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Nr=s(require("graphql-tag")),Wr=Nr.default`
  query storeConfig {
    storeConfig {
        allow_guests_to_write_product_reviews,
        allow_items,
        allow_order,
        base_currency_code,
        base_media_url,
        catalog_default_sort_by,
        category_fixed_product_tax_display_setting,
        cms_home_page,
        cms_no_cookies,
        cms_no_route,
        configurable_thumbnail_source,
        copyright,
        default_description,
        default_display_currency_code,
        default_keywords,
        default_title,
        grid_per_page,
        grid_per_page_values,
        head_shortcut_icon,
        header_logo_src,
        is_default_store,
        is_default_store_group,
        list_mode,
        list_per_page,
        list_per_page_values,
        locale,
        logo_alt,
        logo_height,
        logo_width,
        magento_wishlist_general_is_enabled,
        minimum_password_length,
        no_route,
        product_fixed_product_tax_display_setting,
        product_reviews_enabled,
        required_character_classes_number,
        root_category_uid,
        sales_fixed_product_tax_display_setting,
        store_code,
        store_group_code,
        store_group_name,
        store_name,
        store_sort_order,
        timezone,
        title_prefix,
        title_separator,
        title_suffix,
        use_store_in_url,
        website_code,
        website_name,
        weight_unit,
        welcome,
    }
  }
`;async function oe(e,r={storeConfig:"storeConfig"}){let{storeConfig:o}=e.extendQuery(r,{storeConfig:{query:Wr}});return e.client.query({query:o.query,context:{headers:a(e)}})}var Yr=s(require("graphql-tag")),Jr=Yr.default`
  mutation subscribeEmailToNewsletter($email: String!){
    subscribeEmailToNewsletter(email: $email) {
      status
    }
  }
`;async function ae(e,{email:r},o={subscribeEmailToNewsletter:"subscribeEmailToNewsletter"}){let{subscribeEmailToNewsletter:t}=e.extendQuery(o,{subscribeEmailToNewsletter:{query:Jr,variables:{email:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var jr=s(require("graphql-tag")),Kr=jr.default`
  mutation updateCartItems($input: UpdateCartItemsInput) {
    updateCartItems(input: $input) {
      cart {
        id
        email
        is_virtual
        applied_coupons {
          code
        }
        prices {
          subtotal_excluding_tax {
            value
          },
          subtotal_including_tax {
            value
          },
          applied_taxes {
            amount {
              value
            },
            label
          }
          discounts {
            amount {
              value
            },
            label
          }
          grand_total {
            value
          }
        }
        items {
          uid
          product {
            uid
            __typename
            sku
            name
            stock_status
            only_x_left_in_stock
            rating_summary
            thumbnail {
              url
              position
              disabled
              label
            }
            url_key
            url_rewrites {
              url
            }
            price_range {
              maximum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
              minimum_price {
                final_price {
                  currency
                  value
                }
                regular_price {
                  currency
                  value
                }
              }
            }
            categories {
              uid
              name
              url_suffix
              url_path
              breadcrumbs {
                category_name,
                category_url_path
              }
            }
            review_count
            reviews {
              items {
                average_rating
                ratings_breakdown {
                  name
                  value
                }
              }
            }
          }
          prices {
            row_total {
              value
            }
            row_total_including_tax {
              value
            }
            total_item_discount {
              value
            }
          }
          quantity
          ... on ConfigurableCartItem {
            configurable_options {
              configurable_product_option_uid
              option_label
              configurable_product_option_value_uid
              value_label
            }
            configured_variant {
              sku
              thumbnail {
                url
              }
            }
          }
          ... on BundleCartItem {
            bundle_options {
              uid
              label
              type
              values {
                id
                label
                price
                quantity
              }
            }
          }
        }
        total_quantity
        shipping_addresses {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
          selected_shipping_method {
            carrier_code
            carrier_title
            method_code
            method_title
            amount {
              value
              currency
            }
          }
        }
        billing_address {
          firstname
          lastname
          street
          city
          company
          region {
            code
            region_id
            label
          }
          postcode
          telephone
          country {
            code
            label
          }
        }
      }
    }
  }
`;async function ie(e,r,o={updateCartItems:"updateCartItems"}){let{updateCartItems:t}=e.extendQuery(o,{updateCartItems:{query:Kr,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var Xr=s(require("graphql-tag")),Zr=Xr.default`
  mutation updateCustomer($input: CustomerUpdateInput!) {
    updateCustomerV2(input: $input) {
      customer {
        date_of_birth
        default_billing
        default_shipping
        email
        firstname
        is_subscribed
        lastname
        middlename
        prefix
        suffix
        taxvat
        addresses {
          city
          country_code
          default_billing
          default_shipping
          extension_attributes {
            attribute_code
            value
          }
          firstname
          id
          lastname
          postcode
          prefix
          region {
            region_code
            region_id
            region
          }
          street
          suffix
          telephone
          vat_id
        }
      }
    }
  }
`;var eo=async(e,r,o={updateCustomer:"updateCustomer"})=>{let{updateCustomer:t}=e.extendQuery(o,{updateCustomer:{query:Zr,variables:{input:r}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var to=s(require("graphql-tag")),ro=to.default`
  mutation updateCustomerAddress($id: Int!, $input: CustomerAddressInput) {
    updateCustomerAddress(id: $id, input: $input) {
      id
      city
      company
      country_code
      default_billing
      default_shipping
      extension_attributes {
        attribute_code
        value
      }
      fax
      firstname
      id
      lastname
      middlename
      postcode
      prefix
      region {
        region
        region_code
        region_id
      }
      street
      suffix
      telephone
      vat_id
    }
  }
`;async function ue(e,r,o={updateCustomerAddress:"updateCustomerAddress"}){let{updateCustomerAddress:t}=e.extendQuery(o,{updateCustomerAddress:{query:ro,variables:{id:r.addressId,input:r.input}}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})}var oo=s(require("graphql-tag")),ao=oo.default`
  mutation updateCustomerEmail($email: String!, $password: String!) {
    updateCustomerEmail(email: $email, password: $password){
      customer {
        date_of_birth
        default_billing
        default_shipping
        email
        firstname
        is_subscribed
        lastname
        middlename
        prefix
        suffix
        taxvat
        addresses {
          city
          country_code
          default_billing
          default_shipping
          extension_attributes {
            attribute_code
            value
          }
          firstname
          id
          lastname
          postcode
          prefix
          region {
            region_code
            region_id
            region
          }
          street
          suffix
          telephone
          vat_id
        }
      }
    }
  }
`;var io=async(e,r,o={updateCustomerEmail:"updateCustomerEmail"})=>{let{updateCustomerEmail:t}=e.extendQuery(o,{updateCustomerEmail:{query:ao,variables:n({},r)}});return e.client.mutate({mutation:t.query,variables:t.variables,context:{headers:a(e)}})};var uo=s(require("graphql-tag")),so=uo.default`
  query upsellProducts(
    $search: String = ""
    $filter: ProductAttributeFilterInput
    $pageSize: Int = 10
    $currentPage: Int = 1
    $sort: ProductAttributeSortInput
  ) {
    products(
      search: $search
      filter: $filter
      sort: $sort
      pageSize: $pageSize
      currentPage: $currentPage
    ){
      items {
        upsell_products {
          uid
    __typename
    sku
    name
    stock_status
    only_x_left_in_stock
    rating_summary
    thumbnail {
      url
      position
      disabled
      label
    }
    url_key
    url_rewrites {
      url
    }
    price_range {
      maximum_price {
        final_price {
          currency
          value
        }
        regular_price {
          currency
          value
        }
      }
      minimum_price {
        final_price {
          currency
          value
        }
        regular_price {
          currency
          value
        }
      }
    }
    categories {
      uid
      name
      url_suffix
      url_path
      breadcrumbs {
        category_name,
        category_url_path
      }
    }
    review_count
    reviews {
      items {
        average_rating
        ratings_breakdown {
          name
          value
        }
      }
    }
        }
        uid
      }
    }
  }
`;var no=async(e,r,o={upsellProducts:"upsellProducts"})=>{var l,m;let t=n({pageSize:10,currentPage:1},r),i={pageSize:t.pageSize,currentPage:t.currentPage};t.search&&(i.search=t.search),t.filter&&(i.filter=t.filter),t.sort&&(i.sort=t.sort);let{upsellProducts:u}=e.extendQuery(o,{upsellProducts:{query:so,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}};var lo=s(require("graphql-tag")),mo=lo.default`
  query urlResolver($url: String!) {
    urlResolver(url: $url) {
      id
      redirectCode
      relative_url
      type
      entity_uid
    }
  }
`;async function se(e,r,o={urlResolver:"urlResolver"}){let{urlResolver:t}=e.extendQuery(o,{urlResolver:{query:mo,variables:{url:r}}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var co=s(require("graphql-tag")),po=co.default`
  query route($url: String!) {
    route(url: $url) {
      type
       ... on ProductInterface {
        sku
      }
      ... on CategoryTree {
        uid
      }
      ... on CmsPage {
        identifier
      }
    }
  }
`;async function ne(e,r,o={route:"route"}){let{route:t}=e.extendQuery(o,{route:{query:po,variables:{url:r}}});return e.client.query({query:t.query,variables:t.variables,context:{headers:a(e)}})}var _o=s(require("graphql-tag")),yo=_o.default`
  query wishlist($currentPage: Int = 1, $pageSize: Int = 10) {
    customer {
      wishlists {
        id
        items_count
        sharing_code
        items_v2(currentPage: $currentPage, pageSize: $pageSize) {
          items {
            id
            quantity
            description
            added_at
            product {
              ...on ConfigurableProduct {
                configurable_options {
                  attribute_code
                  attribute_uid
                  label
                  position
                  uid
                  use_default
                  values {
                    label
                    swatch_data {
                      value
                    }
                    uid
                  }
                }
              }
              ... on BundleProduct {
                items {
                  sku
                  title
                  options {
                    uid
                    quantity
                    product {
                      uid
                      sku
                      name
                      price_range {
                        maximum_price {
                          final_price {
                            currency
                            value
                          }
                          regular_price {
                            currency
                            value
                          }
                        }
                        minimum_price {
                          final_price {
                            currency
                            value
                          }
                          regular_price {
                            currency
                            value
                          }
                        }
                      }
                    }
                  }
                }
              }
              uid
              __typename
              sku
              name
              stock_status
              only_x_left_in_stock
              rating_summary
              categories {
                uid
                name
                url_suffix
                url_path
                breadcrumbs {
                  category_name,
                  category_url_path
                }
              }
              price_range {
                maximum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
                minimum_price {
                  final_price {
                    currency
                    value
                  }
                  regular_price {
                    currency
                    value
                  }
                }
              }
              thumbnail {
                url
                position
                disabled
                label
              }
              url_key
              url_rewrites {
                url
              }
              review_count
              reviews {
                items {
                  average_rating
                  ratings_breakdown {
                    name
                    value
                  }
                }
              }
              ... on GroupedProduct {
                items {
                  product {
                    sku
                  }
                }
              }
            }
          }
          page_info {
            current_page
            page_size
            total_pages
          }
        }
      }
    }
  }
`;var fo=async(e,r,o={wishlist:"wishlist"})=>{var l,m;let t=n({pageSize:10,currentPage:1},r),i={pageSize:t.pageSize<=0?10:t.pageSize,currentPage:t.currentPage<=0?1:t.currentPage},{wishlist:u}=e.extendQuery(o,{wishlist:{query:yo,variables:i}});try{return await e.client.query({query:u.query,variables:u.variables,context:{headers:a(e)}})}catch(d){throw((l=d.graphQLErrors)==null?void 0:l[0].message)||((m=d.networkError)==null?void 0:m.result)||d}};var go=s(require("graphql-tag")),Co=go.default`
  query wishlist {
    customer {
      wishlists {
        items_count
      }
    }
  }
`;var bo=async(e,r={wishlistItemsCount:"wishlistItemsCount"})=>{var t,i;let{wishlistItemsCount:o}=e.extendQuery(r,{wishlistItemsCount:{query:Co}});try{return await e.client.query({query:o.query,context:{headers:a(e)}})}catch(u){throw((t=u.graphQLErrors)==null?void 0:t[0].message)||((i=u.networkError)==null?void 0:i.result)||u}};var To=require("@vue-storefront/core");var g=require("@apollo/client/core"),Io=s(require("isomorphic-fetch")),P=require("@vue-storefront/core"),Ao=require("@apollo/client/link/error"),ko=require("@apollo/client/link/retry"),Mo=require("@apollo/client/link/context"),Ro=s(require("agentkeepalive"));var b=require("@vue-storefront/core"),ho=require("@apollo/client/link/context"),vo=()=>(e,r,o)=>{var t;return e>3?!1:((t=o==null?void 0:o.result)==null?void 0:t.message)==="invalid_token"?(b.Logger.debug(`Apollo retry-link, the operation (${r.operationName}) sent with wrong token, creating a new one... (attempt: ${e})`),!0):!1},Po=({state:e})=>(0,ho.setContext)((r,{headers:o})=>{b.Logger.debug("Apollo linkFactory",r.operationName);let t=e.getStore(),i=e.getCustomerToken(),u=e.getCurrency();return u&&b.Logger.debug("Apollo currencyLinkFactory, finished, currency: ",u),t&&b.Logger.debug("Apollo storeLinkFactory, finished, storeId: ",t),i&&b.Logger.debug("Apollo authLinkFactory, finished, token: ",i),{headers:n(n(n(n({},o),u?{"Content-Currency":u}:{}),i?{authorization:`Bearer ${i}`}:{}),t?{Store:t}:{})}});var Uo=["BillingCartAddress","ShippingCartAddress"],No=["SimpleCartItem","VirtualCartItem","DownloadableCartItem","BundleCartItem","ConfigurableCartItem"],Wo=["VirtualProduct","SimpleProduct","DownloadableProduct","BundleProduct","GroupedProduct","ConfigurableProduct"],Yo=["CategoryTree"],Jo=["ProductImage","ProductVideo"],jo=["ProductLinks"],Ko=["CategoryTree","CmsPage","VirtualProduct","SimpleProduct","DownloadableProduct","BundleProduct","GroupedProduct","ConfigurableProduct"],Xo=["DownloadableCreditMemoItem","BundleCreditMemoItem","CreditMemoItem"],Zo=["DownloadableOrderItem","BundleOrderItem","OrderItem"],ea=["DownloadableInvoiceItem","BundleInvoiceItem","InvoiceItem"],ta=["BundleShipmentItem","ShipmentItem"],ra=["SimpleWishlistItem","VirtualWishlistItem","DownloadableWishlistItem","BundleWishlistItem","GroupedProductWishlistItem","ConfigurableWishlistItem"],oa=["AggregationOption"],aa=["LayerFilterItem","SwatchLayerFilterItem"],ia=["SimpleProduct","BundleProduct","GroupedProduct","ConfigurableProduct"],ua=["CustomizableAreaOption","CustomizableDateOption","CustomizableDropDownOption","CustomizableMultipleOption","CustomizableFieldOption","CustomizableFileOption","CustomizableRadioOption","CustomizableCheckboxOption"],sa=["VirtualProduct","SimpleProduct","DownloadableProduct","BundleProduct","ConfigurableProduct"],na=["ImageSwatchData","TextSwatchData","ColorSwatchData"],la=["SwatchLayerFilterItem"],xo={CartAddressInterface:Uo,CartItemInterface:No,ProductInterface:Wo,CategoryInterface:Yo,MediaGalleryInterface:Jo,ProductLinksInterface:jo,RoutableInterface:Ko,CreditMemoItemInterface:Xo,OrderItemInterface:Zo,InvoiceItemInterface:ea,ShipmentItemInterface:ta,WishlistItemInterface:ra,AggregationOptionInterface:oa,LayerFilterItemInterface:aa,PhysicalProductInterface:ia,CustomizableOptionInterface:ua,CustomizableProductInterface:sa,SwatchDataInterface:na,SwatchLayerFilterItemInterface:la};var Qo=require("graphql"),wo=require("node:url"),qo=e=>{let r=new wo.URL(e),o=r.searchParams.get("query");return o?(r.searchParams.set("query",(0,Qo.stripIgnoredCharacters)(o)),r.toString()):e};var{HttpsAgent:ma}=Ro.default,da=new ma,ca=()=>(0,Ao.onError)(({graphQLErrors:e,networkError:r})=>{e&&e.forEach(({message:o,locations:t,path:i,extensions:u})=>{if((u==null?void 0:u.category)!=="graphql-authorization"&&!o.includes("Resource Owner Password Credentials Grant")){if(!t){P.Logger.error(`[GraphQL error]: Message: ${o}, Path: ${i}`);return}let l=t.map(({column:m,line:d})=>`[column: ${m}, line: ${d}]`);P.Logger.error(`[GraphQL error]: Message: ${o}, Location: ${l.join(", ")}, Path: ${i}`)}}),r&&P.Logger.error(`[Network error]: ${r}`)}),So=(e,r)=>{let o=(r==null?void 0:r.apolloLink)||(0,Mo.setContext)((l,{headers:m})=>({headers:n({},m)})),t=new g.HttpLink(n({uri:e.api,fetch:(l,m)=>(0,Io.default)(qo(l),m),fetchOptions:{agent:da}},e.customApolloHttpLinkOptions)),i=ca(),u=new ko.RetryLink({attempts:vo(),delay:()=>0});return(0,g.from)([i,u,o.concat(t)])},me=e=>new g.ApolloClient(n({cache:new g.InMemoryCache({possibleTypes:xo,resultCaching:!0}),queryDeduplication:!0,ssrMode:!0},e));var $o=e=>(To.Logger.debug("createMagentoConnection"),{apolloLink:So(e,{apolloLink:Po({state:e.state})})});var C={api:"",cookies:{currencyCookieName:"vsf-currency",countryCookieName:"vsf-country",localeCookieName:"vsf-locale",cartCookieName:"vsf-cart",customerCookieName:"vsf-customer",storeCookieName:"vsf-store"},state:{getCartId:()=>"",setCartId:()=>{},getCustomerToken:()=>"",setCustomerToken:()=>{},getStore:()=>"",setStore:()=>{},getCurrency:()=>"",setCurrency:()=>{},getLocale:()=>"",setLocale:()=>{},getCountry:()=>"",setCountry:()=>{},setMessage:()=>{},getMessage:()=>({})},externalCheckout:{enable:!1,syncUrlPath:"/vue/cart/sync",stores:{},cmsUrl:""}};var Eo=e=>y(n(n({},C),e),{state:e.state||C.state}),Oo=e=>{let r=Eo(e);if(e.client)return{client:e.client,config:r};if(e.customOptions&&e.customOptions.link)return{client:me(e.customOptions),config:r};let{apolloLink:o}=$o(r),t=me(y(n({link:o},e.customOptions),{defaultOptions:{query:{errorPolicy:"all",fetchPolicy:"no-cache"},mutate:{errorPolicy:"all"}}}));return{config:r,client:t}},pa=e=>(e==null?void 0:e.client)?{config:Eo(e),client:e.client}:Oo(e),_a={name:"tokenExtension",hooks:(e,r)=>({beforeCreate:({configuration:o})=>{var m,d,de,ce;let t=((m=o.cookies)==null?void 0:m.cartCookieName)||C.cookies.cartCookieName,i=((d=o.cookies)==null?void 0:d.customerCookieName)||C.cookies.customerCookieName,u=((de=o.cookies)==null?void 0:de.storeCookieName)||C.cookies.storeCookieName,l=((ce=o.cookies)==null?void 0:ce.currencyCookieName)||C.cookies.currencyCookieName;return y(n({},o),{state:{getCartId:()=>e.cookies[t],setCartId:p=>{if(!p){delete e.cookies[t];return}r.cookie(t,JSON.stringify(p))},getCustomerToken:()=>e.cookies[i],setCustomerToken:p=>{if(!p){delete e.cookies[i];return}r.cookie(i,JSON.stringify(p))},getStore:()=>e.cookies[u],setStore:p=>{if(!p){delete e.cookies[u];return}r.cookie(u,JSON.stringify(p))},getCurrency:()=>e.cookies[l],setCurrency:p=>{if(!p){delete e.cookies[l];return}r.cookie(l,JSON.stringify(p))}}})}})},{createApiClient:ya}=(0,Lo.apiClientFactory)({onCreate:pa,api:le,extensions:[_a]});module.exports=Do(fa);0&&(module.exports={createApiClient,init});
